File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonReaderUtils.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.internal.CodedUtils;
import cn.wjybxx.dson.internal.DsonInternals;
import cn.wjybxx.dson.io.DsonChunk;
import cn.wjybxx.dson.io.DsonIOException;
import cn.wjybxx.dson.io.DsonInput;
import cn.wjybxx.dson.io.DsonOutput;
import cn.wjybxx.dson.types.*;

import java.util.List;

/**
 * @author wjybxx
 * date - 2023/5/31
 */
public class DsonReaderUtils {

    /** 支持读取为bytes和直接写入bytes的数据类型 -- 这些类型不可以存储额外数据在WireType上 */
    public static final List<DsonType> VALUE_BYTES_TYPES = List.of(DsonType.STRING,
            DsonType.BINARY, DsonType.ARRAY, DsonType.OBJECT, DsonType.HEADER);

    // region number

    /**
     * 1.浮点数的前16位固定写入，因此只统计后16位
     * 2.wireType表示后导0对应的字节数
     * 3.由于编码依赖了上层的wireType比特位，因此不能写在Output接口中
     */
    public static int wireTypeOfFloat(float value) {
        int rawBits = Float.floatToRawIntBits(value);
        if ((rawBits & 0xFF) != 0) {
            return 0;
        }
        if ((rawBits & 0xFF00) != 0) {
            return 1;
        }
        return 2;
    }

    /** 小端编码，从末尾非0开始写入 */
    public static void writeFloat(DsonOutput output, float value, int wireType) {
        if (wireType == 0) {
            output.writeFloat(value);
            return;
        }

        int rawBits = Float.floatToRawIntBits(value);
        for (int i = 0; i < wireType; i++) {
            rawBits = rawBits >>> 8;
        }
        for (int i = wireType; i < 4; i++) {
            output.writeRawByte((byte) rawBits);
            rawBits = rawBits >>> 8;
        }
    }

    public static float readFloat(DsonInput input, int wireType) {
        if (wireType == 0) {
            return input.readFloat();
        }

        int rawBits = 0;
        for (int i = wireType; i < 4; i++) {
            rawBits |= (input.readRawByte() & 0XFF) << (8 * i);
        }
        return Float.intBitsToFloat(rawBits);
    }

    /**
     * 浮点数的前16位固定写入，因此只统计后48位
     */
    public static int wireTypeOfDouble(double value) {
        long rawBits = Double.doubleToRawLongBits(value);
        for (int i = 0; i < 6; i++) {
            byte v = (byte) rawBits;
            if (v != 0) {
                return i;
            }
            rawBits = rawBits >>> 8;
        }
        return 6;
    }

    public static void writeDouble(DsonOutput output, double value, int wireType) {
        if (wireType == 0) {
            output.writeDouble(value);
            return;
        }

        long rawBits = Double.doubleToRawLongBits(value);
        for (int i = 0; i < wireType; i++) {
            rawBits = rawBits >>> 8;
        }
        for (int i = wireType; i < 8; i++) {
            output.writeRawByte((byte) rawBits);
            rawBits = rawBits >>> 8;
        }
    }

    public static double readDouble(DsonInput input, int wireType) {
        if (wireType == 0) {
            return input.readDouble();
        }

        long rawBits = 0;
        for (int i = wireType; i < 8; i++) {
            rawBits |= (input.readRawByte() & 0XFFL) << (8 * i);
        }
        return Double.longBitsToDouble(rawBits);
    }

    public static boolean readBool(DsonInput input, int wireTypeBits) {
        if (wireTypeBits == 1) {
            return true;
        }
        if (wireTypeBits == 0) {
            return false;
        }
        throw new DsonIOException("invalid wireType for bool, bits: " + wireTypeBits);
    }
    // endregion

    // region binary

    public static void writeBinary(DsonOutput output, Binary binary) {
        int sizeOfBinaryType = CodedUtils.computeRawVarInt32Size(binary.getType());
        output.writeFixed32(sizeOfBinaryType + binary.getData().length);
        output.writeUint32(binary.getType());
        output.writeRawBytes(binary.getData());
    }

    public static void writeBinary(DsonOutput output, int type, DsonChunk chunk) {
        int sizeOfBinaryType = CodedUtils.computeRawVarInt32Size(type);
        output.writeFixed32(sizeOfBinaryType + chunk.getLength());
        output.writeUint32(type);
        output.writeRawBytes(chunk.getBuffer(), chunk.getOffset(), chunk.getLength());
    }

    public static Binary readBinary(DsonInput input) {
        int size = input.readFixed32();
        int oldLimit = input.pushLimit(size);
        Binary binary;
        {
            int type = input.readUint32();
            int dataLength = input.getBytesUntilLimit();
            binary = new Binary(type, input.readRawBytes(dataLength));
        }
        input.popLimit(oldLimit);
        return binary;
    }

    // endregion

    // region 内置二元组
    public static void writeExtInt32(DsonOutput output, ExtInt32 extInt32, WireType wireType) {
        output.writeUint32(extInt32.getType());
        output.writeBool(extInt32.hasValue());
        if (extInt32.hasValue()) {
            wireType.writeInt32(output, extInt32.getValue());
        }
    }

    public static ExtInt32 readExtInt32(DsonInput input, WireType wireType) {
        int type = input.readUint32();
        boolean hasValue = input.readBool();
        int value = hasValue ? wireType.readInt32(input) : 0;
        return new ExtInt32(type, value, hasValue);
    }

    public static void writeExtInt64(DsonOutput output, ExtInt64 extInt64, WireType wireType) {
        output.writeUint32(extInt64.getType());
        output.writeBool(extInt64.hasValue());
        if (extInt64.hasValue()) {
            wireType.writeInt64(output, extInt64.getValue());
        }
    }

    public static ExtInt64 readExtInt64(DsonInput input, WireType wireType) {
        int type = input.readUint32();
        boolean hasValue = input.readBool();
        long value = hasValue ? wireType.readInt64(input) : 0;
        return new ExtInt64(type, value, hasValue);
    }

    public static void writeExtDouble(DsonOutput output, ExtDouble extDouble, int wireType) {
        output.writeUint32(extDouble.getType());
        output.writeBool(extDouble.hasValue());
        if (extDouble.hasValue()) {
            writeDouble(output, extDouble.getValue(), wireType);
        }
    }

    public static ExtDouble readExtDouble(DsonInput input, int wireType) {
        int type = input.readUint32();
        boolean hasValue = input.readBool();
        double value = hasValue ? readDouble(input, wireType) : 0;
        return new ExtDouble(type, value, hasValue);
    }

    public static int wireTypeOfExtString(ExtString extString) {
        int v = 0;
        if (extString.getType() != 0) {
            v |= ExtString.MASK_TYPE;
        }
        if (extString.hasValue()) {
            v |= ExtString.MASK_VALUE;
        }
        return v;
    }

    public static void writeExtString(DsonOutput output, ExtString extString) {
        if (extString.getType() != 0) {
            output.writeUint32(extString.getType());
        }
        if (extString.hasValue()) {
            output.writeString(extString.getValue());
        }
    }

    public static ExtString readExtString(DsonInput input, int wireTypeBits) {
        int type = DsonInternals.isEnabled(wireTypeBits, ExtString.MASK_TYPE) ? input.readUint32() : 0;
        String value = DsonInternals.isEnabled(wireTypeBits, ExtString.MASK_VALUE) ? input.readString() : null;
        return new ExtString(type, value);
    }
    // endregion

    // region 内置结构体
    public static int wireTypeOfRef(ObjectRef objectRef) {
        int v = 0;
        if (objectRef.hasNamespace()) {
            v |= ObjectRef.MASK_NAMESPACE;
        }
        if (objectRef.getType() != 0) {
            v |= ObjectRef.MASK_TYPE;
        }
        if (objectRef.getPolicy() != 0) {
            v |= ObjectRef.MASK_POLICY;
        }
        return v;
    }

    public static void writeRef(DsonOutput output, ObjectRef objectRef) {
        output.writeString(objectRef.getLocalId());
        if (objectRef.hasNamespace()) {
            output.writeString(objectRef.getNamespace());
        }
        if (objectRef.getType() != 0) {
            output.writeUint32(objectRef.getType());
        }
        if (objectRef.getPolicy() != 0) {
            output.writeUint32(objectRef.getPolicy());
        }
    }

    public static ObjectRef readRef(DsonInput input, int wireTypeBits) {
        String localId = input.readString();
        String namespace = DsonInternals.isEnabled(wireTypeBits, ObjectRef.MASK_NAMESPACE) ? input.readString() : null;
        int type = DsonInternals.isEnabled(wireTypeBits, ObjectRef.MASK_TYPE) ? input.readUint32() : 0;
        int policy = DsonInternals.isEnabled(wireTypeBits, ObjectRef.MASK_POLICY) ? input.readUint32() : 0;
        return new ObjectRef(localId, namespace, type, policy);
    }

    public static void writeTimestamp(DsonOutput output, OffsetTimestamp timestamp) {
        output.writeUint64(timestamp.getSeconds());
        output.writeUint32(timestamp.getNanos());
        output.writeSint32(timestamp.getOffset());
        output.writeUint32(timestamp.getEnables());
    }

    public static OffsetTimestamp readTimestamp(DsonInput input) {
        return new OffsetTimestamp(
                input.readUint64(),
                input.readUint32(),
                input.readSint32(),
                input.readUint32());
    }
    // endregion

    // region 特殊
    public static void writeValueBytes(DsonOutput output, DsonType type, byte[] data) {
        if (type == DsonType.STRING) {
            output.writeUint32(data.length);
        } else {
            output.writeFixed32(data.length);
        }
        output.writeRawBytes(data);
    }

    public static byte[] readValueAsBytes(DsonInput input, DsonType dsonType) {
        int size;
        if (dsonType == DsonType.STRING) {
            size = input.readUint32();
        } else {
            size = input.readFixed32();
        }
        return input.readRawBytes(size);
    }

    public static void checkReadValueAsBytes(DsonType dsonType) {
        if (!VALUE_BYTES_TYPES.contains(dsonType)) {
            throw DsonIOException.invalidDsonType(VALUE_BYTES_TYPES, dsonType);
        }
    }

    public static void checkWriteValueAsBytes(DsonType dsonType) {
        if (!VALUE_BYTES_TYPES.contains(dsonType)) {
            throw DsonIOException.invalidDsonType(VALUE_BYTES_TYPES, dsonType);
        }
    }

    public static void skipToEndOfObject(DsonInput input) {
        int size = input.getBytesUntilLimit();
        if (size > 0) {
            input.skipRawBytes(size);
        }
    }
    // endregion

    public static void skipValue(DsonInput input, DsonContextType contextType,
                                 DsonType dsonType, WireType wireType, int wireTypeBits) {
        int skip;
        switch (dsonType) {
            case FLOAT -> {
                skip = 4 - wireTypeBits;
            }
            case DOUBLE -> {
                skip = 8 - wireTypeBits;
            }
            case BOOLEAN, NULL -> {
                return;
            }
            case INT32 -> {
                wireType.readInt32(input);
                return;
            }
            case INT64 -> {
                wireType.readInt64(input);
                return;
            }
            case STRING -> {
                skip = input.readUint32();  // string长度
            }
            case EXT_INT32 -> {
                input.readUint32(); // 子类型
                if (input.readBool()) {
                    wireType.readInt32(input);
                }
                return;
            }
            case EXT_INT64 -> {
                input.readUint32(); // 子类型
                if (input.readBool()) {
                    wireType.readInt64(input);
                }
                return;
            }
            case EXT_DOUBLE -> {
                input.readUint32(); // 子类型
                skip = input.readBool() ? (8 - wireTypeBits) : 0;
                input.skipRawBytes(skip);
                return;
            }
            case EXT_STRING -> {
                if (DsonInternals.isEnabled(wireTypeBits, ExtString.MASK_TYPE)) {
                    input.readUint32(); // 子类型
                }
                if (DsonInternals.isEnabled(wireTypeBits, ExtString.MASK_VALUE)) {
                    skip = input.readUint32(); // string长度
                    input.skipRawBytes(skip);
                }
                return;
            }
            case REFERENCE -> {
                skip = input.readUint32(); // localId
                input.skipRawBytes(skip);

                if (DsonInternals.isEnabled(wireTypeBits, ObjectRef.MASK_NAMESPACE)) {
                    skip = input.readUint32(); // namespace
                    input.skipRawBytes(skip);
                }
                if (DsonInternals.isEnabled(wireTypeBits, ObjectRef.MASK_TYPE)) {
                    input.readUint32();
                }
                if (DsonInternals.isEnabled(wireTypeBits, ObjectRef.MASK_POLICY)) {
                    input.readUint32();
                }
                return;
            }
            case TIMESTAMP -> {
                input.readUint64();
                input.readUint32();
                input.readSint32();
                input.readUint32();
                return;
            }
            case BINARY, ARRAY, OBJECT, HEADER -> {
                skip = input.readFixed32();
            }
            default -> {
                throw DsonIOException.invalidDsonType(contextType, dsonType);
            }
        }
        if (skip > 0) {
            input.skipRawBytes(skip);
        }
    }

    public static DsonReaderGuide whatShouldIDo(DsonContextType contextType, DsonReaderState state) {
        if (contextType == DsonContextType.TOP_LEVEL) {
            if (state == DsonReaderState.END_OF_FILE) {
                return DsonReaderGuide.CLOSE;
            }
            if (state == DsonReaderState.VALUE) {
                return DsonReaderGuide.READ_VALUE;
            }
            return DsonReaderGuide.READ_TYPE;
        } else {
            return switch (state) {
                case TYPE -> DsonReaderGuide.READ_TYPE;
                case VALUE -> DsonReaderGuide.READ_VALUE;
                case NAME -> DsonReaderGuide.READ_NAME;
                case WAIT_START_OBJECT -> {
                    if (contextType == DsonContextType.HEADER) {
                        yield DsonReaderGuide.START_HEADER;
                    }
                    if (contextType == DsonContextType.ARRAY) {
                        yield DsonReaderGuide.START_ARRAY;
                    }
                    yield DsonReaderGuide.START_OBJECT;
                }
                case WAIT_END_OBJECT -> {
                    if (contextType == DsonContextType.HEADER) {
                        yield DsonReaderGuide.END_HEADER;
                    }
                    if (contextType == DsonContextType.ARRAY) {
                        yield DsonReaderGuide.END_ARRAY;
                    }
                    yield DsonReaderGuide.END_OBJECT;
                }
                case INITIAL, END_OF_FILE -> throw new AssertionError("invalid state " + state);
            };
        }
    }

}

File: csharp/Dson/src/DsonReaderUtils.cs
#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using Wjybxx.Dson.IO;
using Wjybxx.Dson.Types;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// Dson二进制编解码工具类
/// </summary>
public static class DsonReaderUtils
{
    /** 支持读取为bytes和直接写入bytes的数据类型 -- 这些类型不可以存储额外数据在WireType上 */
    public static readonly IList<DsonType> ValueBytesTypes = new[]
    {
        DsonType.String, DsonType.Binary, DsonType.Array, DsonType.Object, DsonType.Header
    }.ToImmutableList();

    #region number

    public static void WriteInt32(IDsonOutput output, int value, WireType wireType) {
        switch (wireType) {
            case WireType.VarInt: {
                output.WriteInt32(value);
                break;
            }
            case WireType.Uint: {
                output.WriteUint32(value);
                break;
            }
            case WireType.Sint: {
                output.WriteSint32(value);
                break;
            }
            case WireType.Fixed: {
                output.WriteFixed32(value);
                break;
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(wireType), wireType, null);
        }
    }

    public static int ReadInt32(IDsonInput input, WireType wireType) {
        switch (wireType) {
            case WireType.VarInt: {
                return input.ReadInt32();
            }
            case WireType.Uint: {
                return input.ReadUint32();
            }
            case WireType.Sint: {
                return input.ReadSint32();
            }
            case WireType.Fixed: {
                return input.ReadFixed32();
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(wireType), wireType, null);
        }
    }

    public static void WriteInt64(IDsonOutput output, long value, WireType wireType) {
        switch (wireType) {
            case WireType.VarInt: {
                output.WriteInt64(value);
                break;
            }
            case WireType.Uint: {
                output.WriteUint64(value);
                break;
            }
            case WireType.Sint: {
                output.WriteSint64(value);
                break;
            }
            case WireType.Fixed: {
                output.WriteFixed64(value);
                break;
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(wireType), wireType, null);
        }
    }

    public static long ReadInt64(IDsonInput input, WireType wireType) {
        switch (wireType) {
            case WireType.VarInt: {
                return input.ReadInt64();
            }
            case WireType.Uint: {
                return input.ReadUint64();
            }
            case WireType.Sint: {
                return input.ReadSint64();
            }
            case WireType.Fixed: {
                return input.ReadFixed64();
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(wireType), wireType, null);
        }
    }

    /**
     * 1.浮点数的前16位固定写入，因此只统计后16位
     * 2.wireType表示后导0对应的字节数
     * 3.由于编码依赖了上层的wireType比特位，因此不能写在Output接口中
     */
    public static int WireTypeOfFloat(float value) {
        int rawBits = BitConverter.SingleToInt32Bits(value);
        if ((rawBits & 0xFF) != 0) {
            return 0;
        }
        if ((rawBits & 0xFF00) != 0) {
            return 1;
        }
        return 2;
    }

    /** 小端编码，从末尾非0开始写入 */
    public static void WriteFloat(IDsonOutput output, float value, int wireType) {
        if (wireType == 0) {
            output.WriteFloat(value);
            return;
        }

        int rawBits = BitConverter.SingleToInt32Bits(value);
        for (int i = 0; i < wireType; i++) {
            rawBits = rawBits >> 8;
        }
        for (int i = wireType; i < 4; i++) {
            output.WriteRawByte((byte)rawBits);
            rawBits = rawBits >> 8;
        }
    }

    public static float ReadFloat(IDsonInput input, int wireType) {
        if (wireType == 0) {
            return input.ReadFloat();
        }

        int rawBits = 0;
        for (int i = wireType; i < 4; i++) {
            rawBits |= (input.ReadRawByte() & 0XFF) << (8 * i);
        }
        return BitConverter.Int32BitsToSingle(rawBits);
    }

    /**
     * 浮点数的前16位固定写入，因此只统计后48位
     */
    public static int WireTypeOfDouble(double value) {
        long rawBits = BitConverter.DoubleToInt64Bits(value);
        for (int i = 0; i < 6; i++) {
            byte v = (byte)rawBits;
            if (v != 0) {
                return i;
            }
            rawBits = rawBits >> 8;
        }
        return 6;
    }

    public static void WriteDouble(IDsonOutput output, double value, int wireType) {
        if (wireType == 0) {
            output.WriteDouble(value);
            return;
        }

        long rawBits = BitConverter.DoubleToInt64Bits(value);
        for (int i = 0; i < wireType; i++) {
            rawBits = rawBits >> 8;
        }
        for (int i = wireType; i < 8; i++) {
            output.WriteRawByte((byte)rawBits);
            rawBits = rawBits >> 8;
        }
    }

    public static double ReadDouble(IDsonInput input, int wireType) {
        if (wireType == 0) {
            return input.ReadDouble();
        }

        long rawBits = 0;
        for (int i = wireType; i < 8; i++) {
            rawBits |= (input.ReadRawByte() & 0XFFL) << (8 * i);
        }
        return BitConverter.Int64BitsToDouble(rawBits);
    }

    public static bool ReadBool(IDsonInput input, int wireTypeBits) {
        if (wireTypeBits == 1) {
            return true;
        }
        if (wireTypeBits == 0) {
            return false;
        }
        throw new DsonIOException("invalid wireType for bool, bits: " + wireTypeBits);
    }

    #endregion

    #region binary

    public static void WriteBinary(IDsonOutput output, Binary binary) {
        int sizeOfBinaryType = BinaryUtils.ComputeRawVarInt32Size((uint)binary.Type);
        output.WriteFixed32(sizeOfBinaryType + binary.Data.Length);
        output.WriteUint32(binary.Type);
        output.WriteRawBytes(binary.Data);
    }

    public static void WriteBinary(IDsonOutput output, int type, DsonChunk chunk) {
        int sizeOfBinaryType = BinaryUtils.ComputeRawVarInt32Size((uint)type);
        output.WriteFixed32(sizeOfBinaryType + chunk.Length);
        output.WriteUint32(type);
        output.WriteRawBytes(chunk.Buffer, chunk.Offset, chunk.Length);
    }

    public static Binary ReadBinary(IDsonInput input) {
        int size = input.ReadFixed32();
        int oldLimit = input.PushLimit(size);
        Binary binary;
        {
            int type = input.ReadUint32();
            int dataLength = input.GetBytesUntilLimit();
            binary = new Binary(type, input.ReadRawBytes(dataLength));
        }
        input.PopLimit(oldLimit);
        return binary;
    }

    #endregion

    #region 内置二元组

    public static void WriteExtInt32(IDsonOutput output, ExtInt32 extInt32, WireType wireType) {
        output.WriteUint32(extInt32.Type);
        output.WriteBool(extInt32.HasValue);
        if (extInt32.HasValue) {
            WriteInt32(output, extInt32.Value, wireType);
        }
    }

    public static ExtInt32 ReadExtInt32(IDsonInput input, WireType wireType) {
        int type = input.ReadUint32();
        bool hasValue = input.ReadBool();
        int value = hasValue ? ReadInt32(input, wireType) : 0;
        return new ExtInt32(type, value, hasValue);
    }

    public static void WriteExtInt64(IDsonOutput output, ExtInt64 extInt64, WireType wireType) {
        output.WriteUint32(extInt64.Type);
        output.WriteBool(extInt64.HasValue);
        if (extInt64.HasValue) {
            WriteInt64(output, extInt64.Value, wireType);
        }
    }

    public static ExtInt64 ReadExtInt64(IDsonInput input, WireType wireType) {
        int type = input.ReadUint32();
        bool hasValue = input.ReadBool();
        long value = hasValue ? ReadInt64(input, wireType) : 0;
        return new ExtInt64(type, value, hasValue);
    }

    public static void WriteExtDouble(IDsonOutput output, ExtDouble extDouble, int wireType) {
        output.WriteUint32(extDouble.Type);
        output.WriteBool(extDouble.HasValue);
        if (extDouble.HasValue) {
            WriteDouble(output, extDouble.Value, wireType);
        }
    }

    public static ExtDouble ReadExtDouble(IDsonInput input, int wireType) {
        int type = input.ReadUint32();
        bool hasValue = input.ReadBool();
        double value = hasValue ? ReadDouble(input, wireType) : 0;
        return new ExtDouble(type, value, hasValue);
    }

    public static int WireTypeOfExtString(ExtString extString) {
        int v = 0;
        if (extString.Type != 0) {
            v |= ExtString.MaskType;
        }
        if (extString.HasValue) {
            v |= ExtString.MaskValue;
        }
        return v;
    }

    public static void WriteExtString(IDsonOutput output, ExtString extString) {
        if (extString.Type != 0) {
            output.WriteUint32(extString.Type);
        }
        if (extString.HasValue) {
            output.WriteString(extString.Value!);
        }
    }

    public static ExtString ReadExtString(IDsonInput input, int wireTypeBits) {
        int type = DsonInternals.IsEnabled(wireTypeBits, ExtString.MaskType) ? input.ReadUint32() : 0;
        string value = DsonInternals.IsEnabled(wireTypeBits, ExtString.MaskValue) ? input.ReadString() : null;
        return new ExtString(type, value);
    }

    #endregion

    #region 内置结构体

    public static int WireTypeOfRef(ObjectRef objectRef) {
        int v = 0;
        if (objectRef.hasNamespace) {
            v |= ObjectRef.MaskNamespace;
        }
        if (objectRef.Type != 0) {
            v |= ObjectRef.MaskType;
        }
        if (objectRef.Policy != 0) {
            v |= ObjectRef.MaskPolicy;
        }
        return v;
    }

    public static void WriteRef(IDsonOutput output, ObjectRef objectRef) {
        output.WriteString(objectRef.LocalId);
        if (objectRef.hasNamespace) {
            output.WriteString(objectRef.Ns);
        }
        if (objectRef.Type != 0) {
            output.WriteUint32(objectRef.Type);
        }
        if (objectRef.Policy != 0) {
            output.WriteUint32(objectRef.Policy);
        }
    }

    public static ObjectRef ReadRef(IDsonInput input, int wireTypeBits) {
        string localId = input.ReadString();
        string ns = DsonInternals.IsEnabled(wireTypeBits, ObjectRef.MaskNamespace) ? input.ReadString() : null;
        int type = DsonInternals.IsEnabled(wireTypeBits, ObjectRef.MaskType) ? input.ReadUint32() : 0;
        int policy = DsonInternals.IsEnabled(wireTypeBits, ObjectRef.MaskPolicy) ? input.ReadUint32() : 0;
        return new ObjectRef(localId, ns, type, policy);
    }

    public static void WriteTimestamp(IDsonOutput output, OffsetTimestamp timestamp) {
        output.WriteUint64(timestamp.Seconds);
        output.WriteUint32(timestamp.Nanos);
        output.WriteSint32(timestamp.Offset);
        output.WriteUint32(timestamp.Enables);
    }

    public static OffsetTimestamp ReadTimestamp(IDsonInput input) {
        return new OffsetTimestamp(
            input.ReadUint64(),
            input.ReadUint32(),
            input.ReadSint32(),
            input.ReadUint32());
    }

    #endregion

    #region 特殊

    public static void WriteValueBytes(IDsonOutput output, DsonType type, byte[] data) {
        if (type == DsonType.String) {
            output.WriteUint32(data.Length);
        } else {
            output.WriteFixed32(data.Length);
        }
        output.WriteRawBytes(data);
    }

    public static byte[] ReadValueAsBytes(IDsonInput input, DsonType dsonType) {
        int size;
        if (dsonType == DsonType.String) {
            size = input.ReadUint32();
        } else {
            size = input.ReadFixed32();
        }
        return input.ReadRawBytes(size);
    }

    public static void CheckReadValueAsBytes(DsonType dsonType) {
        if (!ValueBytesTypes.Contains(dsonType)) {
            throw DsonIOException.InvalidDsonType(ValueBytesTypes, dsonType);
        }
    }

    public static void CheckWriteValueAsBytes(DsonType dsonType) {
        if (!ValueBytesTypes.Contains(dsonType)) {
            throw DsonIOException.InvalidDsonType(ValueBytesTypes, dsonType);
        }
    }

    public static void SkipToEndOfObject(IDsonInput input) {
        int size = input.GetBytesUntilLimit();
        if (size > 0) {
            input.SkipRawBytes(size);
        }
    }

    #endregion

    public static void SkipValue(IDsonInput input, DsonContextType contextType,
                                 DsonType dsonType, WireType wireType, int wireTypeBits) {
        int skip;
        switch (dsonType) {
            case DsonType.Float: {
                skip = 4 - wireTypeBits;
                break;
            }
            case DsonType.Double: {
                skip = 8 - wireTypeBits;
                break;
            }
            case DsonType.Boolean:
            case DsonType.Null: {
                return;
            }
            case DsonType.Int32: {
                ReadInt32(input, wireType);
                return;
            }
            case DsonType.Int64: {
                ReadInt64(input, wireType);
                return;
            }
            case DsonType.String: {
                skip = input.ReadUint32(); // string长度
                break;
            }
            case DsonType.ExtInt32: {
                input.ReadUint32(); // 子类型
                if (input.ReadBool()) {
                    ReadInt32(input, wireType);
                }
                return;
            }
            case DsonType.ExtInt64: {
                input.ReadUint32(); // 子类型
                if (input.ReadBool()) {
                    ReadInt64(input, wireType);
                }
                return;
            }
            case DsonType.ExtDouble: {
                input.ReadUint32(); // 子类型
                skip = input.ReadBool() ? (8 - wireTypeBits) : 0;
                input.SkipRawBytes(skip);
                return;
            }
            case DsonType.ExtString: {
                if (DsonInternals.IsEnabled(wireTypeBits, ExtString.MaskType)) {
                    input.ReadUint32(); // 子类型
                }
                if (DsonInternals.IsEnabled(wireTypeBits, ExtString.MaskValue)) {
                    skip = input.ReadUint32(); // string长度
                    input.SkipRawBytes(skip);
                }
                return;
            }
            case DsonType.Reference: {
                skip = input.ReadUint32(); // localId
                input.SkipRawBytes(skip);

                if (DsonInternals.IsEnabled(wireTypeBits, ObjectRef.MaskNamespace)) {
                    skip = input.ReadUint32(); // namespace
                    input.SkipRawBytes(skip);
                }
                if (DsonInternals.IsEnabled(wireTypeBits, ObjectRef.MaskType)) {
                    input.ReadUint32();
                }
                if (DsonInternals.IsEnabled(wireTypeBits, ObjectRef.MaskPolicy)) {
                    input.ReadUint32();
                }
                return;
            }
            case DsonType.Timestamp: {
                input.ReadUint64();
                input.ReadUint32();
                input.ReadSint32();
                input.ReadUint32();
                return;
            }
            case DsonType.Binary:
            case DsonType.Array:
            case DsonType.Object:
            case DsonType.Header: {
                skip = input.ReadFixed32();
                break;
            }
            default: {
                throw DsonIOException.InvalidDsonType(contextType, dsonType);
            }
        }
        if (skip > 0) {
            input.SkipRawBytes(skip);
        }
    }

    public static DsonReaderGuide WhatShouldIDo(DsonContextType contextType, DsonReaderState state) {
        if (contextType == DsonContextType.TopLevel) {
            if (state == DsonReaderState.EndOfFile) {
                return DsonReaderGuide.Close;
            }
            if (state == DsonReaderState.Value) {
                return DsonReaderGuide.ReadValue;
            }
            return DsonReaderGuide.ReadType;
        }
        switch (state) {
            case DsonReaderState.Type: return DsonReaderGuide.ReadType;
            case DsonReaderState.Value: return DsonReaderGuide.ReadValue;
            case DsonReaderState.Name: return DsonReaderGuide.ReadName;
            case DsonReaderState.WaitStartObject: {
                if (contextType == DsonContextType.Header) {
                    return DsonReaderGuide.StartHeader;
                }
                if (contextType == DsonContextType.Array) {
                    return DsonReaderGuide.StartArray;
                }
                return DsonReaderGuide.StartObject;
            }
            case DsonReaderState.WaitEndObject: {
                if (contextType == DsonContextType.Header) {
                    return DsonReaderGuide.EndHeader;
                }
                if (contextType == DsonContextType.Array) {
                    return DsonReaderGuide.EndArray;
                }
                return DsonReaderGuide.EndObject;
            }
            case DsonReaderState.Initial:
            case DsonReaderState.EndOfFile:
            default:
                throw new InvalidOperationException("invalid state " + state);
        }
    }
}